src/odemis/acq/align/autofocus.py

```diff
@@ -587,10 +591,258 @@ def estimateAutoFocusTime(detector, scanner=None, steps=MAX_STEPS_NUMBER):
     """
     return steps * estimateAcquisitionTime(detector, scanner)
 
+def totalAutofocusTime(spgr, dets):
+    ngs = len(spgr.axes["grating"].choices)
+    nds = len(dets)
+    et = estimateAutoFocusTime(dets[0], None) + 20
+
+    # 1 time for each grating/detector combination
+    move_et = ngs * 20 if ngs > 1 else 0  # extra 20 s for grating moves
+    move_et += nds * 5 if nds > 1 else 0  # extra 5 s for detector selector moves
+
+    return (ngs * nds) * et + move_et
+
+def Sparc2AutoFocus(streams, align_mode, opm, start_autofocus):
+    """
+    Wrapper for _DoSparc2AutoFocus. It provides the ability to check the progress of Sparc2AutoFocus
+    procedure or even cancel it.
+
+    returns (model.ProgressiveFuture):  Progress of _DoSparc2AutoFocus, whose result() will return:
+            Focus position (m)
+            Focus level
+    """
+    # first calculate the focuser, detectors, selectors in order to make the time estimation for the AutoFocusSpectrometer
+    focuser = None
+    # All the streams in the list have the same focuser
+    if isinstance(streams, (list,)):
+        focuser = streams[0].focuser
+        if focuser is None:
+            logging.warning("Stream %s used for focusing has no focuser", streams[0])
+    if focuser is None:
+        focuser = model.getComponent(role='focus')
+        logging.debug("Focuser is None so get component with role focus %s", focuser)
+
+    # Get all the detectors affected by the focuser
+    try:
+        spgr, dets, selector = _getSpectrometerFocusingComponents(focuser)  # type: (object, List[Any], Optional[Any])
+    except LookupError as ex:
+        logging.error("Failed to focus: %s", ex)
+        # TODO: just run the standard autofocus procedure instead?
+        return
+
+    # Create ProgressiveFuture and update its state to RUNNING
+    est_start = time.time() + 0.1
+
+    # Rough approximation of the times of each action:
+    # * 5 s to turn on the light
+    # * 5 s to close the slit
+    # * af_time s for the AutoFocusSpectrometer procedure to be completed
+    # * 5 s to open the slit
```

Comment:
This step doesn't open the slit, but just acquires one last image. You could use the `stream.estimateAcquisitionTime()` (with the first stream), and fall back to 0 if no stream. Alternatively, you could just put 0.2 s, which should be correct most of the time.

Reply:
I put 0.2s, in case of using stream.estimateAcquisitionTime() I am not sure which emitter I should use.